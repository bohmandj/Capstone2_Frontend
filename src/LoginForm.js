import React, { useState, useContext } from "react";
import { Navigate } from "react-router-dom";
import MemoLedgerContext from "./MemoLedgerContext";
import {
    Button,
    Card,
    CardBody,
    CardTitle,
    Form,
    Input,
    Label,
    Nav
} from "reactstrap";

const LoginForm = () => {
    /* Form to take username & password for sign in. */

    const { currentUser, login } = useContext(MemoLedgerContext);

    const inputs = ["username", "password"];

    const INITIAL_STATE = inputs.reduce((obj, input) => {
        obj[input] = "";
        return obj;
    }, {});

    const [formData, setFormData] = useState(INITIAL_STATE);

    const handleChange = (e) => {
        const { name, value } = e.target;
        setFormData(prev => ({
            ...prev,
            [name]: value
        }));
    };

    const handleSubmit = async (e) => {
        e.preventDefault();
        const responses = Object.values(formData);
        const isFormComplete = responses.every((response, i) => {
            return response.trim() !== "";
        });
        if (isFormComplete) {
            login(formData);
        } else {
            alert("Valid username & password must both be provided to log in.")
        };
    };

    if (currentUser) {
        return <Navigate to={'/'} />
    };

    return (
        <div className='form-container container col-md-6 col-lg-5 col-xl-4 mx-auto'>
            <Card>
                <CardBody>
                    <CardTitle tag="h3">
                        Log In
                    </CardTitle>
                    <Form onSubmit={handleSubmit}>
                        {inputs.map(input => (
                            <div className='mb-3' key={input}>
                                <Label
                                    className='text-capitalize font-weight-bold'
                                    htmlFor={input}
                                >{input}</Label>
                                <Input
                                    id={input}
                                    type="text"
                                    name={input}
                                    value={formData[input] || ""}
                                    onChange={handleChange}
                                />
                            </div>
                        ))}
                        <Button name="submit" className='w-100'>Submit</Button>
                    </Form>
                </CardBody>
            </Card>
        </div>
    )
}

export default LoginForm;