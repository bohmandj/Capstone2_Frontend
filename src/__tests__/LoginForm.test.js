import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { BrowserRouter, MemoryRouter } from 'react-router-dom'; // Import MemoryRouter for isolated routing
import LoginForm from '../LoginForm';
import MemoLedgerContext from '../MemoLedgerContext';

// Mock the MemoLedgerContext values that LoginForm uses
const mockLogin = jest.fn();
const mockSetIsLoading = jest.fn(); // Mock setIsLoading as it's used by login indirectly
const mockSetCurrentUser = jest.fn(); // Mock setCurrentUser if needed by other components using the context

const mockContextValueLoggedIn = {
    currentUser: { username: 'testuser' }, // Simulate logged-in user
    login: mockLogin,
    setIsLoading: mockSetIsLoading,
    setCurrentUser: mockSetCurrentUser,
};

const mockContextValueLoggedOut = {
    currentUser: null, // Simulate logged-out user
    login: mockLogin,
    setIsLoading: mockSetIsLoading,
    setCurrentUser: mockSetCurrentUser,
};

// Helper function to render LoginForm within the context and router
const renderLoginForm = (contextValue, initialEntries = ['/login']) => {
    return render(
        <MemoLedgerContext.Provider value={contextValue}>
            <MemoryRouter initialEntries={initialEntries}>
                <LoginForm />
            </MemoryRouter>
        </MemoLedgerContext.Provider>
    );
};

describe('LoginForm', () => {
    // Clear mocks before each test to ensure isolation
    beforeEach(() => {
        mockLogin.mockClear();
        mockSetIsLoading.mockClear();
        mockSetCurrentUser.mockClear();
        jest.spyOn(window, 'alert').mockImplementation(() => { }); // Mock window.alert
    });

    afterEach(() => {
        jest.restoreAllMocks(); // Restore window.alert
    });

    // Smoke Test
    // Renders without crashing when logged out
    test('renders without crashing when logged out', () => {
        renderLoginForm(mockContextValueLoggedOut);
        expect(screen.getByText('Log In')).toBeInTheDocument();
        expect(screen.getByLabelText(/username/i)).toBeInTheDocument();
        expect(screen.getByLabelText(/password/i)).toBeInTheDocument();
    });

    // Renders without crashing when logged in (should redirect)
    test('renders without crashing when logged in and redirects', () => {
        const { asFragment } = renderLoginForm(mockContextValueLoggedIn);
        expect(asFragment()).toMatchSnapshot();
    });

    // Initial Render (Logged Out State)
    test('displays correct initial elements when logged out', () => {
        renderLoginForm(mockContextValueLoggedOut);

        // Check for title
        expect(screen.getByRole('heading', { name: /log in/i, level: 3 })).toBeInTheDocument();

        // Check for input fields
        expect(screen.getByLabelText(/username/i)).toHaveValue('');
        expect(screen.getByLabelText(/password/i)).toHaveValue('');

        // Check for submit button
        expect(screen.getByRole('button', { name: /submit/i })).toBeInTheDocument();
    });

    // Successful Login
    test('calls login function with correct data on successful submission', async () => {
        renderLoginForm(mockContextValueLoggedOut);

        const usernameInput = screen.getByLabelText(/username/i);
        const passwordInput = screen.getByLabelText(/password/i);
        const submitButton = screen.getByRole('button', { name: /submit/i });

        // Simulate user input
        fireEvent.change(usernameInput, { target: { name: 'username', value: 'testuser' } });
        fireEvent.change(passwordInput, { target: { name: 'password', value: 'password123' } });

        // Simulate form submission
        fireEvent.click(submitButton);

        // Assert that login was called with the correct form data
        await waitFor(() => {
            expect(mockLogin).toHaveBeenCalledTimes(1);
            expect(mockLogin).toHaveBeenCalledWith({
                username: 'testuser',
                password: 'password123',
            });
        });
    });

    // Incomplete Form Submission (Missing Username)
    test('shows alert and does NOT call login if username is missing', async () => {
        renderLoginForm(mockContextValueLoggedOut);

        const passwordInput = screen.getByLabelText(/password/i);
        const submitButton = screen.getByRole('button', { name: /submit/i });

        fireEvent.change(passwordInput, { target: { name: 'password', value: 'password123' } });
        fireEvent.click(submitButton);

        // Assert that an alert was shown
        expect(window.alert).toHaveBeenCalledTimes(1);
        expect(window.alert).toHaveBeenCalledWith("Valid username & password must both be provided to log in.");

        // Assert that login was NOT called
        expect(mockLogin).not.toHaveBeenCalled();
    });

    // Incomplete Form Submission (Missing Password)
    test('shows alert and does NOT call login if password is missing', async () => {
        renderLoginForm(mockContextValueLoggedOut);

        const usernameInput = screen.getByLabelText(/username/i);
        const submitButton = screen.getByRole('button', { name: /submit/i });

        fireEvent.change(usernameInput, { target: { name: 'username', value: 'testuser' } });
        fireEvent.click(submitButton);

        // Assert that an alert was shown
        expect(window.alert).toHaveBeenCalledTimes(1);
        expect(window.alert).toHaveBeenCalledWith("Valid username & password must both be provided to log in.");

        // Assert that login was NOT called
        expect(mockLogin).not.toHaveBeenCalled();
    });

    // Snapshot Test
    test('matches snapshot when logged out', () => {
        const { asFragment } = renderLoginForm(mockContextValueLoggedOut);
        expect(asFragment()).toMatchSnapshot();
    });
});