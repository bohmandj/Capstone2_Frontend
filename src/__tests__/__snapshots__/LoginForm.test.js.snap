// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm matches snapshot when logged out 1`] = `
<DocumentFragment>
  <div
    class="page-content form-container container col-11 col-md-6 col-lg-5 col-xl-4 mx-auto my-auto"
  >
    <div
      class="card"
    >
      <div
        class="card-body"
      >
        <h3
          class="card-title"
        >
          Log In
        </h3>
        <form
          class=""
        >
          <div
            class="mb-3"
          >
            <label
              class="text-capitalize font-weight-bold form-label"
              for="username"
            >
              username
            </label>
            <input
              class="form-control"
              id="username"
              name="username"
              type="text"
              value=""
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="text-capitalize font-weight-bold form-label"
              for="password"
            >
              password
            </label>
            <input
              class="form-control"
              id="password"
              name="password"
              type="text"
              value=""
            />
          </div>
          <button
            class="w-100 btn btn-secondary"
            name="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`LoginForm renders without crashing when logged in and redirects 1`] = `<DocumentFragment />`;
